packages/liveblocks-client/src/immutable.ts

```diff
@@ -271,25 +271,32 @@ function getParentsPath(node: LiveNode): Array<string | number> {
   return path;
 }
 
-export function patchImmutableObject<S extends JsonObject>(
+//
+// TODO: Remove `patchImmutableObject`!
+//
+// This helper is now only used internally, to support our Zustand and
+// Redux packages. We should be able to reimplement those using the new
+// `.toImmutable()` APIs.
+//
+export function legacy_patchImmutableObject<S extends JsonObject>(
```

Comment:
One thing we should consider for the zustand/redux integration is whether to keep the incoming immutable reference or always use the one generated from the room's storage.

For example, with a zustand store like this one:

```ts
const useStore = create(
  middleware<State, Presence>(
    (set) => ({
      obj: { a: 0 },
      otherObj: { b: 0 },

      setObj: (newObj) => set({ obj: newObj }),
      setOtherObj: (otherObj) => set({ otherObj: newOtherObj }),
    }),
    {
      client,
      storageMapping: { obj: true, otherObj: true },
    }
  )
);
```

On 0.17, after calling `setObj`, `state.obj` === `newObj`.

If we update the middleware to always synchronize the store with the underlying immutable references, `state.obj` !== `newObj`. It's hard to predict if it will have any impact... it depends on how developers use their store, so I don't think we should take that risk. 

Ideally, we should have a way to replace the immutable cache of a `LiveStructure` without breaking the encapsulation. My first intuition is to add a method on `LiveStructure` that would take an immutable object to update them and replace the immutable cache simultaneously. Basically, moving the code of `patchLiveObject` to `LiveStructure`. What do you think?

Reply:
I've codified this concern in an explicit unit test for this case: see f8a4127827b4c8d24b1100fa778f055a9d59d5a2.